templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,185 @@
+    if (Test-Path $temp_puppetres) {
+      Out-File -InputObject "Restoring puppetres.dll" -FilePath $Logfile -Append
+      Move-Item -Path $temp_puppetres -Destination "$InstallDir\puppet\bin\puppetres.dll"
+    }
+  }
+}
+
+# ************** Execution start **************
+$ErrorActionPreference = "Stop"
+$service_names=@(
+  "puppet",
+  "pxp-agent",
+  "mcollective"
+)
+$msi_arguments = "/qn /norestart /i `"$Source`" /l*vx+ `"$Logfile`" $InstallArgs"
+
+if (Test-Path "HKLM:\SOFTWARE\Puppet Labs") {
+  $msi_arguments += " ENVIRONMENT=`"$(puppet.bat config print --section agent environment)`""
+}
+if ($InstallDir) {
+  $msi_arguments += " INSTALLDIR=`"$InstallDir`""
+}
+if ($PuppetMaster) {
+  $msi_arguments += " PUPPET_MASTER_SERVER=`"$PuppetMaster`""
+}
+if ($PuppetStartType) {
+  $msi_arguments += " PUPPET_AGENT_STARTUP_MODE=`"$PuppetStartType`""
+}
+Out-File -InputObject $PID -FilePath $InstallScriptPIDFile
+try {
+  # Wait for the puppet run to finish
+  Out-File -InputObject "Waiting for puppet to stop, PID:$PuppetPID" -FilePath $Logfile -Append
+  try {
+    Wait-Process -Id $PuppetPID -Timeout 120
+  # We have to catch all exceptions here since the Error action preference of 'stop' will cause wait-process
+  # to actually throw [System.Management.Automation.ActionPreferenceStopException] rather than the actual
+  # exceptions. However, once inside the catch block the value of $_ is correctly set as the actual exception
+  } catch {
+    if ($_.Exception.GetType() -eq [Microsoft.PowerShell.Commands.ProcessCommandException]) {
+      Out-File -InputObject "Puppet was already finished, continuing..." -FilePath $Logfile -Append
+    } elseif ($_.Exception.GetType() -eq [System.TimeoutException]) {
+      Out-File -InputObject "ERROR: timed out waiting for puppet run to finish!" -FilePath $Logfile -Append
+      throw $_
+    } else {
+      Out-File -InputObject "ERROR: failed while trying to wait for puppet process with $($_.Exception.GetType())" -FilePath $Logfile -Append
+      throw $_
+    }
+  }
+  foreach($service in $service_names) {
+    $serv_exists = Get-Service -Name $service -ErrorAction SilentlyContinue
+    if ($serv_exists) {
+      Out-File -InputObject "Stopping $($service) before upgrade" -FilePath $Logfile -Append
+      Stop-Service $service
+    }
+  }
+  if ($UseLockedFilesWorkaround) {
+    $temp_puppetres = move_puppetres_dll
+  }
+  Out-File -InputObject "Beginning MSI installation with Arguments: $msi_arguments" -FilePath $Logfile -Append
+  Out-File -InputObject "****************************** Begin msiexec.exe output ******************************" -FilePath $Logfile -Append
+  $msiexec_proc = Start-Process -FilePath msiexec.exe -PassThru -ArgumentList $msi_arguments
+  # Using the -Wait option in Start-Process will wait on the process group instead of just the one msiexec process in older versions of windows.
+  # but using WaitForExit works fine.
+  $msiexec_proc.WaitForExit()

```

Comment:
Would it be useful to log the MSI's exit code?

Reply:
And also waiting on a timeout?

Reply:
Get's tricky here, what value would you choose for a timeout?

What do you do IF it times out? Just wait longer?

TBH I don't think this requires one.

Reply:
Not sure about the value, I'm not familiar with how long a typical agent install/upgrade takes on Windows. It may be useful to log something if the install's taking longer than average, such as the PID of the MSI process?

I agree that a timeout isn't strictly necessary.

Reply:
And that's kind of the problem.  We don't have any statistics to base a number on :-(

Reply:
I tend to agree that we can't assume any kind of reasonable timeout here. I can guess, but we have no idea what versions of windows and how many resources the windows machine will have on installation.